File: packages/sdk/typescript/human-protocol-sdk/src/kvstore.ts
import {
  KVStore,
  KVStore__factory,
} from '@human-protocol/core/typechain-types';
import { ContractRunner, Overrides, ethers } from 'ethers';
import { BaseEthersClient } from './base';
import { KVStoreKeys, NETWORKS } from './constants';
import { requiresSigner } from './decorators';
import { ChainId } from './enums';
import {
  ErrorInvalidAddress,
  ErrorInvalidHash,
  ErrorInvalidUrl,
  ErrorKVStoreArrayLength,
  ErrorKVStoreEmptyKey,
  ErrorProviderDoesNotExist,
  ErrorUnsupportedChainID,
  InvalidKeyError,
} from './error';
import { NetworkData } from './types';
import { getSubgraphUrl, customGqlFetch, isValidUrl } from './utils';
import {
  GET_KVSTORE_BY_ADDRESS_AND_KEY_QUERY,
  GET_KVSTORE_BY_ADDRESS_QUERY,
} from './graphql/queries/kvstore';
import { KVStoreData } from './graphql';
import { IKVStore, SubgraphOptions } from './interfaces';
/**
 * ## Introduction
 *
 * This client enables performing actions on KVStore contract and obtaining information from both the contracts and subgraph.
 *
 * Internally, the SDK will use one network or another according to the network ID of the `runner`.
 * To use this client, it is recommended to initialize it using the static `build` method.
 *
 * ```ts
 * static async build(runner: ContractRunner): Promise<KVStoreClient>;
 * ```
 *
 * A `Signer` or a `Provider` should be passed depending on the use case of this module:
 *
 * - **Signer**: when the user wants to use this model to send transactions calling the contract functions.
 * - **Provider**: when the user wants to use this model to get information from the contracts or subgraph.
 *
 * ## Installation
 *
 * ### npm
 * ```bash
 * npm install @human-protocol/sdk
 * ```
 *
 * ### yarn
 * ```bash
 * yarn install @human-protocol/sdk
 * ```
 *
 * ## Code example
 *
 * ### Signer
 *
 * **Using private key (backend)**
 *
 * ```ts
 * import { KVStoreClient } from '@human-protocol/sdk';
 * import { Wallet, JsonRpcProvider } from 'ethers';
 *
 * const rpcUrl = 'YOUR_RPC_URL';
 * const privateKey = 'YOUR_PRIVATE_KEY';
 *
 * const provider = new JsonRpcProvider(rpcUrl);
 * const signer = new Wallet(privateKey, provider);
 * const kvstoreClient = await KVStoreClient.build(signer);
 * ```
 *
 * **Using Wagmi (frontend)**
 *
 * ```ts
 * import { useSigner, useChainId } from 'wagmi';
 * import { KVStoreClient } from '@human-protocol/sdk';
 *
 * const { data: signer } = useSigner();
 * const kvstoreClient = await KVStoreClient.build(signer);
 * ```
 *
 * ### Provider
 *
 * ```ts
 * import { KVStoreClient } from '@human-protocol/sdk';
 * import { JsonRpcProvider } from 'ethers';
 *
 * const rpcUrl = 'YOUR_RPC_URL';
 *
 * const provider = new JsonRpcProvider(rpcUrl);
 * const kvstoreClient = await KVStoreClient.build(provider);
 * ```
 */

export class KVStoreClient extends BaseEthersClient {
  private contract: KVStore;

  /**
   * **KVStoreClient constructor**
   *
   * @param runner - The Runner object to interact with the Ethereum network
   * @param networkData - The network information required to connect to the KVStore contract
   */
  constructor(runner: ContractRunner, networkData: NetworkData) {
    super(runner, networkData);

    this.contract = KVStore__factory.connect(
      networkData.kvstoreAddress,
      runner
    );
  }

  /**
   * Creates an instance of KVStoreClient from a runner.
   *
   * @param runner - The Runner object to interact with the Ethereum network
   * @returns An instance of KVStoreClient
   * @throws ErrorProviderDoesNotExist If the provider does not exist for the provided Signer
   * @throws ErrorUnsupportedChainID If the network's chainId is not supported
   *
   * @example
   * ```ts
   * import { KVStoreClient } from '@human-protocol/sdk';
   * import { Wallet, JsonRpcProvider } from 'ethers';
   *
   * const rpcUrl = 'YOUR_RPC_URL';
   * const privateKey = 'YOUR_PRIVATE_KEY';
   *
   * const provider = new JsonRpcProvider(rpcUrl);
   * const signer = new Wallet(privateKey, provider);
   * const kvstoreClient = await KVStoreClient.build(signer);
   * ```
   */
  public static async build(runner: ContractRunner): Promise<KVStoreClient> {
    if (!runner.provider) {
      throw ErrorProviderDoesNotExist;
    }

    const network = await runner.provider?.getNetwork();

    const chainId: ChainId = Number(network?.chainId);
    const networkData = NETWORKS[chainId];

    if (!networkData) {
      throw ErrorUnsupportedChainID;
    }

    return new KVStoreClient(runner, networkData);
  }

  /**
   * This function sets a key-value pair associated with the address that submits the transaction.
   *
   * @param key - Key of the key-value pair
   * @param value - Value of the key-value pair
   * @param txOptions - Additional transaction parameters (optional, defaults to an empty object).
   * @throws ErrorKVStoreEmptyKey If the key is empty
   * @throws Error If the transaction fails
   *
   * @example
   * ```ts
   * await kvstoreClient.set('Role', 'RecordingOracle');
   * ```
   */
  @requiresSigner
  public async set(
    key: string,
    value: string,
    txOptions: Overrides = {}
  ): Promise<void> {
    if (key === '') throw ErrorKVStoreEmptyKey;
    try {
      await (await this.contract.set(key, value, txOptions)).wait();
    } catch (e) {
      if (e instanceof Error) throw Error(`Failed to set value: ${e.message}`);
    }
  }

  /**
   * This function sets key-value pairs in bulk associated with the address that submits the transaction.
   *
   * @param keys - Array of keys (keys and value must have the same order)
   * @param values - Array of values
   * @param txOptions - Additional transaction parameters (optional, defaults to an empty object).
   * @throws ErrorKVStoreArrayLength If keys and values arrays have different lengths
   * @throws ErrorKVStoreEmptyKey If any key is empty
   * @throws Error If the transaction fails
   *
   * @example
   * ```ts
   * const keys = ['role', 'webhook_url'];
   * const values = ['RecordingOracle', 'http://localhost'];
   * await kvstoreClient.setBulk(keys, values);
   * ```
   */
  @requiresSigner
  public async setBulk(
    keys: string[],
    values: string[],
    txOptions: Overrides = {}
  ): Promise<void> {
    if (keys.length !== values.length) throw ErrorKVStoreArrayLength;
    if (keys.includes('')) throw ErrorKVStoreEmptyKey;

    try {
      await (await this.contract.setBulk(keys, values, txOptions)).wait();
    } catch (e) {
      if (e instanceof Error)
        throw Error(`Failed to set bulk values: ${e.message}`);
    }
  }

  /**
   * Sets a URL value for the address that submits the transaction, and its hash.
   *
   * @param url - URL to set
   * @param urlKey - Configurable URL key. `url` by default.
   * @param txOptions - Additional transaction parameters (optional, defaults to an empty object).
   * @throws ErrorInvalidUrl If the URL is invalid
   * @throws Error If the transaction fails
   *
   * @example
   * ```ts
   * await kvstoreClient.setFileUrlAndHash('example.com');
   * await kvstoreClient.setFileUrlAndHash('linkedin.com/example', 'linkedin_url');
   * ```
   */
  @requiresSigner
  public async setFileUrlAndHash(
    url: string,
    urlKey = 'url',
    txOptions: Overrides = {}
  ): Promise<void> {
    if (!isValidUrl(url)) {
      throw ErrorInvalidUrl;
    }

    const content = await fetch(url).then((res) => res.text());
    const contentHash = ethers.keccak256(ethers.toUtf8Bytes(content));

    const hashKey = urlKey + '_hash';

    try {
      await (
        await this.contract.setBulk(
          [urlKey, hashKey],
          [url, contentHash],
          txOptions
        )
      ).wait();
    } catch (e) {
      if (e instanceof Error)
        throw Error(`Failed to set URL and hash: ${e.message}`);
    }
  }
  /**
   * Gets the value of a key-value pair in the contract.
   *
   * @param address - Address from which to get the key value.
   * @param key - Key to obtain the value.
   * @returns Value of the key.
   * @throws ErrorKVStoreEmptyKey If the key is empty
   * @throws ErrorInvalidAddress If the address is invalid
   * @throws Error If the contract call fails
   *
   * @example
   * ```ts
   * const value = await kvstoreClient.get('0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266', 'Role');
   * console.log('Value:', value);
   * ```
   */
  public async get(address: string, key: string): Promise<string> {
    if (key === '') throw ErrorKVStoreEmptyKey;
    if (!ethers.isAddress(address)) throw ErrorInvalidAddress;

    try {
      const result = await this.contract?.get(address, key);
      return result;
    } catch (e) {
      if (e instanceof Error) throw Error(`Failed to get value: ${e.message}`);
      return e;
    }
  }
}

/**
 * Utility class for KVStore-related operations.
 *
 * @example
 * ```ts
 * import { ChainId, KVStoreUtils } from '@human-protocol/sdk';
 *
 * const kvStoreData = await KVStoreUtils.getKVStoreData(
 *   ChainId.POLYGON_AMOY,
 *   "0x1234567890123456789012345678901234567890"
 * );
 * console.log('KVStore data:', kvStoreData);
 * ```
 */
export class KVStoreUtils {
  /**
   * This function returns the KVStore data for a given address.
   *
   * @param chainId - Network in which the KVStore is deployed
   * @param address - Address of the KVStore
   * @param options - Optional configuration for subgraph requests.
   * @returns KVStore data
   * @throws ErrorUnsupportedChainID If the network's chainId is not supported
   * @throws ErrorInvalidAddress If the address is invalid
   *
   * @example
   * ```ts
   * const kvStoreData = await KVStoreUtils.getKVStoreData(
   *   ChainId.POLYGON_AMOY,
   *   "0x1234567890123456789012345678901234567890"
   * );
   * console.log('KVStore data:', kvStoreData);
   * ```
   */
  public static async getKVStoreData(
    chainId: ChainId,
    address: string,
    options?: SubgraphOptions
  ): Promise<IKVStore[]> {
    const networkData = NETWORKS[chainId];

    if (!networkData) {
      throw ErrorUnsupportedChainID;
    }

    if (address && !ethers.isAddress(address)) {
      throw ErrorInvalidAddress;
    }

    const { kvstores } = await customGqlFetch<{ kvstores: KVStoreData[] }>(
      getSubgraphUrl(networkData),
      GET_KVSTORE_BY_ADDRESS_QUERY(),
      { address: address.toLowerCase() },
      options
    );

    const kvStoreData = kvstores.map((item) => ({
      key: item.key,
      value: item.value,
    }));

    return kvStoreData || [];
  }

  /**
   * Gets the value of a key-value pair in the KVStore using the subgraph.
   *
   * @param chainId - Network in which the KVStore is deployed
   * @param address - Address from which to get the key value.
   * @param key - Key to obtain the value.
   * @param options - Optional configuration for subgraph requests.
   * @returns Value of the key.
   * @throws ErrorUnsupportedChainID If the network's chainId is not supported
   * @throws ErrorInvalidAddress If the address is invalid
   * @throws ErrorKVStoreEmptyKey If the key is empty
   * @throws InvalidKeyError If the key is not found
   *
   * @example
   * ```ts
   * const value = await KVStoreUtils.get(
   *   ChainId.POLYGON_AMOY,
   *   '0x1234567890123456789012345678901234567890',
   *   'role'
   * );
   * console.log('Value:', value);
   * ```
   */
  public static async get(
    chainId: ChainId,
    address: string,
    key: string,
    options?: SubgraphOptions
  ): Promise<string> {
    if (key === '') throw ErrorKVStoreEmptyKey;
    if (!ethers.isAddress(address)) throw ErrorInvalidAddress;

    const networkData = NETWORKS[chainId];

    if (!networkData) {
      throw ErrorUnsupportedChainID;
    }

    const { kvstores } = await customGqlFetch<{ kvstores: KVStoreData[] }>(
      getSubgraphUrl(networkData),
      GET_KVSTORE_BY_ADDRESS_AND_KEY_QUERY(),
      { address: address.toLowerCase(), key },
      options
    );

    if (!kvstores || kvstores.length === 0) {
      throw new InvalidKeyError(key, address);
    }

    return kvstores[0].value;
  }

  /**
   * Gets the URL value of the given entity, and verifies its hash.
   *
   * @param chainId - Network in which the KVStore is deployed
   * @param address - Address from which to get the URL value.
   * @param urlKey - Configurable URL key. `url` by default.
   * @param options - Optional configuration for subgraph requests.
   * @returns URL value for the given address if it exists, and the content is valid
   * @throws ErrorInvalidAddress If the address is invalid
   * @throws ErrorInvalidHash If the hash verification fails
   * @throws Error If fetching URL or hash fails
   *
   * @example
   * ```ts
   * const url = await KVStoreUtils.getFileUrlAndVerifyHash(
   *   ChainId.POLYGON_AMOY,
   *   '0x1234567890123456789012345678901234567890'
   * );
   * console.log('Verified URL:', url);
   * ```
   */
  public static async getFileUrlAndVerifyHash(
    chainId: ChainId,
    address: string,
    urlKey = 'url',
    options?: SubgraphOptions
  ): Promise<string> {
    if (!ethers.isAddress(address)) throw ErrorInvalidAddress;
    const hashKey = urlKey + '_hash';

    let url = '',
      hash = '';

    try {
      url = await this.get(chainId, address, urlKey, options);
    } catch (e) {
      if (e instanceof Error) throw Error(`Failed to get URL: ${e.message}`);
    }

    // Return empty string
    if (!url?.length) {
      return '';
    }

    try {
      hash = await this.get(chainId, address, hashKey);
    } catch (e) {
      if (e instanceof Error) throw Error(`Failed to get Hash: ${e.message}`);
    }

    const content = await fetch(url).then((res) => res.text());
    const contentHash = ethers.keccak256(ethers.toUtf8Bytes(content));

    const formattedHash = hash?.replace(/^0x/, '');
    const formattedContentHash = contentHash?.replace(/^0x/, '');

    if (formattedHash !== formattedContentHash) {
      throw ErrorInvalidHash;
    }

    return url;
  }

  /**
   * Gets the public key of the given entity, and verifies its hash.
   *
   * @param chainId - Network in which the KVStore is deployed
   * @param address - Address from which to get the public key.
   * @param options - Optional configuration for subgraph requests.
   * @returns Public key for the given address if it exists, and the content is valid
   * @throws ErrorInvalidAddress If the address is invalid
   * @throws ErrorInvalidHash If the hash verification fails
   * @throws Error If fetching the public key fails
   *
   * @example
   * ```ts
   * const publicKey = await KVStoreUtils.getPublicKey(
   *   ChainId.POLYGON_AMOY,
   *   '0x1234567890123456789012345678901234567890'
   * );
   * console.log('Public key:', publicKey);
   * ```
   */
  public static async getPublicKey(
    chainId: ChainId,
    address: string,
    options?: SubgraphOptions
  ): Promise<string> {
    const publicKeyUrl = await this.getFileUrlAndVerifyHash(
      chainId,
      address,
      KVStoreKeys.publicKey,
      options
    );

    if (publicKeyUrl === '') {
      return '';
    }

    const publicKey = await fetch(publicKeyUrl).then((res) => res.text());

    return publicKey;
  }
}

File: packages/sdk/rust/human-protocol-sdk/src/kvstore.rs
use web3::{
    signing::{ SecretKey, SecretKeyRef, keccak256 },
    contract::{Contract, Options},
    transports::Http,
    types::Address
};

use crate::enums::EthereumError;
use crate::constants::CONFIRMATIONS_THRESHOLD;
use url::Url;
use hex::encode;


pub struct KVStore {
    contract: Contract<Http>
}

impl KVStore {
    fn new(web3: &web3::Web3<web3::transports::Http>, address: Address) -> Self {
        let contract: Contract<Http> = Contract::from_json(
                web3.eth(),
                address,
                include_bytes!("../../../../core/abis/KVStore.json")
            )
            .unwrap();
        
        KVStore {
            contract
        }
    }
}

/// # Introduction
///
/// A client for interacting with the KVStore contract on the blockchain.
/// 
pub struct KVStoreClient<'a> {
    web3: &'a web3::Web3<web3::transports::Http>,
    kvstore: KVStore,
    account: SecretKey
}

impl<'a> KVStoreClient<'a> {
    /// Creates a new `KVStoreClient` instance.
    ///
    /// # Arguments
    ///
    /// * `web3` - A reference to the `Web3` instance for interacting with the blockchain.
    /// * `address` - The Ethereum address of the KVStore contract.
    /// * `account` - The Ethereum account used for signing transactions.
    ///
    /// # Returns
    ///
    /// Returns a `KVStoreClient` instance.
    ///
    /// # Example
    ///
    /// ```rust
    /// # use std::str::FromStr;
    /// # use human_protocol_sdk::{kvstore::{KVStoreClient}, enums::EthereumError};
    /// # use web3::transports::Http;
    /// # use web3::{signing::{ SecretKey, SecretKeyRef }, types::Address};
    /// 
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), EthereumError> {
    /// let http = Http::new("https://mainnet.infura.io/v3/YOUR_INFURA_API_KEY").unwrap();
    /// let web3 = web3::Web3::new(http);
    /// let address = Address::from_str("0x1234567890123456789012345678901234567890").unwrap();
    /// let account = SecretKey::from_slice(&[1; 32]).unwrap();
    ///
    /// let kvstore_client = KVStoreClient::new(&web3, address, account).await;
    /// #     Ok(())
    /// # }
    /// ```
    pub async fn new(web3: &'a web3::Web3<web3::transports::Http>, address: Address, account: SecretKey) -> KVStoreClient<'a> {
        let kvstore = KVStore::new(web3, address);
        
        KVStoreClient {
            web3,
            kvstore,
            account
        }
    }
    
    /// Sets a key-value pair in the KVStore contract on the blockchain.
    ///
    /// # Arguments
    ///
    /// * `key` - The key for the key-value pair.
    /// * `value` - The value associated with the key.
    ///
    /// # Errors
    ///
    /// Returns an `EthereumError` if the transaction fails or encounters an error.
    ///
    /// # Example
    ///
    /// ```rust
    /// # use std::str::FromStr;
    /// # use human_protocol_sdk::{kvstore::{KVStoreClient}, enums::EthereumError};
    /// # use web3::transports::Http;
    /// # use web3::{signing::{ SecretKey, SecretKeyRef }, types::Address};
    /// 
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), EthereumError> {
    /// # let http = Http::new("https://mainnet.infura.io/v3/YOUR_INFURA_API_KEY").unwrap();
    /// # let web3 = web3::Web3::new(http);
    /// # let address = Address::from_str("0x1234567890123456789012345678901234567890").unwrap();
    /// # let account = SecretKey::from_slice(&[1; 32]).unwrap();
    /// # let kvstore_client = KVStoreClient::new(&web3, address, account).await;
    ///
    /// let key = "example_key".to_string();
    /// let value = "example_value".to_string();
    ///
    /// match kvstore_client.set(
    ///     key,
    ///     value
    /// ).await {
    ///     Ok(_) => println!("Value has been set"),
    ///     Err(err) => println!("{:?}", err)
    /// };
    /// #     Ok(())
    /// # }
    /// ```
    pub async fn set(
        &self,
        key: String, 
        value: String
    ) -> Result<(), EthereumError> {
        let result = self
            .kvstore
            .contract
            .signed_call_with_confirmations(
                "set",
                (
                    key,
                    value,
                ), 
                Options {
                    gas: Some(5_000_000.into()),
                    ..Default::default()
                },
                CONFIRMATIONS_THRESHOLD.into(),
                SecretKeyRef::new(&self.account)
            )
            .await;

        match result {
            Ok(receipt) => {
                if receipt.status.map_or(false, |status| status == 0.into()) {
                    return Err(EthereumError::TransactionFailedError)
                }
                Ok(())
            }
            Err(err) => return Err(EthereumError::Web3Error(err.to_string()))
        }
    }

    /// Sets multiple key-value pairs in the KVStore contract on the blockchain.
    ///
    /// # Arguments
    ///
    /// * `keys` - A vector of keys for the key-value pairs.
    /// * `values` - A vector of values associated with the keys.
    ///
    /// # Errors
    ///
    /// Returns an `EthereumError` if the transaction fails or encounters an error.
    ///
    /// # Example
    /// 
    /// ```rust
    /// # use std::str::FromStr;
    /// # use human_protocol_sdk::{kvstore::{KVStoreClient}, enums::EthereumError};
    /// # use web3::transports::Http;
    /// # use web3::{signing::{ SecretKey, SecretKeyRef }, types::Address};
    /// 
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), EthereumError> {
    /// # let http = Http::new("https://mainnet.infura.io/v3/YOUR_INFURA_API_KEY").unwrap();
    /// # let web3 = web3::Web3::new(http);
    /// # let address = Address::from_str("0x1234567890123456789012345678901234567890").unwrap();
    /// # let account = SecretKey::from_slice(&[1; 32]).unwrap();
    /// # let kvstore_client = KVStoreClient::new(&web3, address, account).await;
    ///
    /// let keys = vec!["key1".to_string(), "key2".to_string()];
    /// let values = vec!["value1".to_string(), "value2".to_string()];
    ///
    /// match kvstore_client.set_bulk(
    ///     keys,
    ///     values
    /// ).await {
    ///     Ok(_) => println!("Values have been set"),
    ///     Err(err) => println!("{:?}", err)
    /// };
    /// #     Ok(())
    /// # }
    /// ```
    pub async fn set_bulk(
        &self,
        keys: Vec<String>, 
        values: Vec<String>
    ) -> Result<(), EthereumError> {
        if keys.len() == 0 {
            return Err(EthereumError::InvalidKeysSizeError)
        }

        if values.len() == 0 {
            return Err(EthereumError::InvalidValuesSizeError)
        }

        if keys.len() != values.len() {
            return Err(EthereumError::RecipientsAndAmountsMustBeSameLengthError)
        }
        
        let result = self
            .kvstore
            .contract
            .signed_call_with_confirmations(
                "set",
                (
                    keys,
                    values,
                ), 
                Options {
                    gas: Some(5_000_000.into()),
                    ..Default::default()
                },
                CONFIRMATIONS_THRESHOLD.into(),
                SecretKeyRef::new(&self.account)
            )
            .await;

        match result {
            Ok(receipt) => {
                if receipt.status.map_or(false, |status| status == 0.into()) {
                    return Err(EthereumError::TransactionFailedError)
                }
                Ok(())
            }
            Err(err) => return Err(EthereumError::Web3Error(err.to_string()))
        }
    }

    /// Sets a key-value pair in the KVStore contract on the blockchain, where the value is fetched from a specified URL.
    ///
    /// # Arguments
    ///
    /// * `url` - The URL from which to fetch the value.
    /// * `url_key` - The key associated with the fetched value.
    ///
    /// # Errors
    ///
    /// Returns an `EthereumError` if the transaction fails, encounters an error, or if the URL is invalid.
    ///
    /// # Example
    ///
    /// ```rust
    /// # use std::str::FromStr;
    /// # use human_protocol_sdk::{kvstore::{KVStoreClient}, enums::EthereumError};
    /// # use web3::transports::Http;
    /// # use web3::{signing::{ SecretKey, SecretKeyRef }, types::Address};
    /// 
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), EthereumError> {
    /// # let http = Http::new("https://mainnet.infura.io/v3/YOUR_INFURA_API_KEY").unwrap();
    /// # let web3 = web3::Web3::new(http);
    /// # let address = Address::from_str("0x1234567890123456789012345678901234567890").unwrap();
    /// # let account = SecretKey::from_slice(&[1; 32]).unwrap();
    /// # let kvstore_client = KVStoreClient::new(&web3, address, account).await;
    ///
    /// let url = "https://example.com/value".to_string();
    /// let url_key = "url".to_string();
    ///
    /// match kvstore_client.set_url(
    ///     url,
    ///     url_key
    /// ).await {
    ///     Ok(_) => println!("Url has been set"),
    ///     Err(err) => println!("{:?}", err)
    /// };
    /// #     Ok(())
    /// # }
    /// ```
    pub async fn set_url(
        &self,
        url: String, 
        url_key: String
    ) -> Result<(), EthereumError> {
        if url.is_empty() || !KVStoreClient::is_valid_url(&url) {
            return Err(EthereumError::InvalidUrlError);
        }

        let content: String = KVStoreClient::fetch_content(&url).await?;
        let hash_bytes = keccak256(content.as_bytes());
        let hash_hex = encode(&hash_bytes);

        let hash_key = format!("{}Hash", url_key);

        let result = self
            .kvstore
            .contract
            .signed_call_with_confirmations(
                "set_bulk",
                (
                    [url_key, hash_key], [url, hash_hex]
                ), 
                Options {
                    gas: Some(5_000_000.into()),
                    ..Default::default()
                },
                CONFIRMATIONS_THRESHOLD.into(),
                SecretKeyRef::new(&self.account)
            )
            .await;

        match result {
            Ok(receipt) => {
                if receipt.status.map_or(false, |status| status == 0.into()) {
                    return Err(EthereumError::TransactionFailedError)
                }
                Ok(())
            }
            Err(err) => return Err(EthereumError::Web3Error(err.to_string()))
        }
    }

    /// Gets the value associated with a key from the KVStore contract on the blockchain.
    ///
    /// # Arguments
    ///
    /// * `address` - The address associated with the stored key-value pair.
    /// * `key` - The key for which to retrieve the value.
    ///
    /// # Errors
    ///
    /// Returns an `EthereumError` if the transaction fails, encounters an error, or if the key is invalid.
    ///
    /// # Example
    ///
    /// ```rust
    /// # use std::str::FromStr;
    /// # use human_protocol_sdk::{kvstore::{KVStoreClient}, enums::EthereumError};
    /// # use web3::transports::Http;
    /// # use web3::{signing::{ SecretKey, SecretKeyRef }, types::Address};
    /// 
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), EthereumError> {
    /// # let http = Http::new("https://mainnet.infura.io/v3/YOUR_INFURA_API_KEY").unwrap();
    /// # let web3 = web3::Web3::new(http);
    /// # let address = Address::from_str("0x1234567890123456789012345678901234567890").unwrap();
    /// # let account = SecretKey::from_slice(&[1; 32]).unwrap();
    /// # let kvstore_client = KVStoreClient::new(&web3, address, account).await;
    ///
    /// let key = "example_key".to_string();
    ///
    /// match kvstore_client.get(
    ///     address,
    ///     key
    /// ).await {
    ///     Ok(value) => println!("Stored value {:?}", value),
    ///     Err(err) => println!("{:?}", err)
    /// };
    /// #     Ok(())
    /// # }
    /// ```
    pub async fn get(
        &self,
        address: Address, 
        key: String
    ) -> Result<String, EthereumError> {
        if key.is_empty() {
            return Err(EthereumError::InvalidKeyError);
        }

        let value: Address = self
            .kvstore
            .contract
            .query(
                "get", 
                (
                    address,
                    key
                ), 
                None, 
                Options::default(), 
                None
            )
            .await
            .map_err(|web3_error| EthereumError::Web3ContractError(web3_error.to_string()))?;
        
        
        Ok(format!("{:?}", value))
    }

    /// Gets the URL associated with a key from the KVStore contract on the Ethereum blockchain.
    ///
    /// # Arguments
    ///
    /// * `address` - The address associated with the stored URL.
    /// * `url_key` - The key for which to retrieve the URL.
    ///
    /// # Errors
    ///
    /// Returns an `EthereumError` if the transaction fails, encounters an error, or if the key is invalid.
    ///
    /// # Example
    ///
    /// ```rust
    /// # use std::str::FromStr;
    /// # use human_protocol_sdk::{kvstore::{KVStoreClient}, enums::EthereumError};
    /// # use web3::transports::Http;
    /// # use web3::{signing::{ SecretKey, SecretKeyRef }, types::Address};
    /// 
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), EthereumError> {
    /// # let http = Http::new("https://mainnet.infura.io/v3/YOUR_INFURA_API_KEY").unwrap();
    /// # let web3 = web3::Web3::new(http);
    /// # let address = Address::from_str("0x1234567890123456789012345678901234567890").unwrap();
    /// # let account = SecretKey::from_slice(&[1; 32]).unwrap();
    /// # let kvstore_client = KVStoreClient::new(&web3, address, account).await;
    ///
    /// let url_key = "url".to_string();
    /// 
    /// match kvstore_client.get_url(
    ///     address,
    ///     url_key
    /// ).await {
    ///     Ok(value) => println!("Stored value {:?}", value),
    ///     Err(err) => println!("{:?}", err)
    /// };
    /// 
    /// #     Ok(())
    /// # }
    /// ```
    pub async fn get_url(
        &self,
        address: Address, 
        url_key: String
    ) -> Result<String, EthereumError> {
        if url_key.is_empty() {
            return Err(EthereumError::InvalidKeyError);
        }

        let hash_key = format!("{}Hash", url_key);

        let url: String = self.get(address, url_key).await?;
        /*let url: String = self
            .kvstore
            .contract
            .query(
                "get", 
                (
                    address,
                    url_key
                ), 
                None, 
                Options::default(), 
                None
            )
            .await
            .map_err(|web3_error| EthereumError::Web3ContractError(web3_error.to_string()))?;
*/
        if url.is_empty() || !KVStoreClient::is_valid_url(&url) {
            return Err(EthereumError::InvalidUrlError);
        }
        
        let hash: String = self.get(address, hash_key).await?;
        /*
        let hash: String = self
            .kvstore
            .contract
            .query(
                "get", 
                (
                    address,
                    hash_key
                ), 
                None, 
                Options::default(), 
                None
            )
            .await
            .map_err(|web3_error| EthereumError::Web3ContractError(web3_error.to_string()))?;
*/
        let content: String = KVStoreClient::fetch_content(&url).await?;
        let hash_bytes = keccak256(content.as_bytes());
        let hash_hex = encode(&hash_bytes);
        
        if hash != hash_hex {
            EthereumError::InvalidHashError(hash, hash_hex);
        }
        
        Ok(format!("{:?}", url))
    }

    /// Checks if the provided string is a valid URL.
    ///
    /// # Arguments
    ///
    /// * `url_str` - The string representing the URL to check for validity.
    ///
    /// # Returns
    ///
    /// Returns `true` if the provided string is a valid URL, otherwise returns `false`.
    fn is_valid_url(url_str: &str) -> bool {
        match Url::parse(url_str) {
            Ok(_) => true,
            Err(_) => false,
        }
    }

    /// Fetches content from the specified URL using an HTTP GET request.
    ///
    /// # Arguments
    ///
    /// * `url` - The URL from which to fetch the content.
    ///
    /// # Returns
    ///
    /// Returns a `Result` containing the fetched content as a `String` if the request is successful, otherwise returns an `Err` with the specific error.
    ///
    /// # Errors
    ///
    /// The function may return the following errors:
    ///
    /// * `RequestError` - An error occurred while making the HTTP request.
    /// * `UnexpectedStatusCodeError` - The HTTP request returned an unexpected status code.
    /// ```
    async fn fetch_content(url: &str) -> Result<String, EthereumError> {
        let client = reqwest::Client::new();
    
        let response = client.get(url).send().await.map_err(|err| EthereumError::RequestError(err.to_string()))?;
    
        match response.status() {
            reqwest::StatusCode::OK => {
                let content = response.text().await.map_err(|err| EthereumError::RequestError(err.to_string()))?;
                Ok(content)
            },
            status => Err(EthereumError::UnexpectedStatusCodeError(status.as_u16())),
        }
    }
}

#[cfg(test)]
mod tests {}